### ISSUE TITLE
Error when you try to lock a locked door from Siri

### ISSUE THREAD

**Issue by kodsnutten** [kodsnutten/homebridge-verisure#2]
As reported by @siavashvj, Homebridge is restarted after an error, when you try to lock a already locked door, or unlock an unlocked door...

From the log:
`Error: This callback function has already been called by someone else; it can only be called one time.
at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/once.js:12:13
at Object. (/usr/local/lib/node_modules/homebridge-verisure/index.js:293:11)
at Request._callback (/usr/local/lib/node_modules/homebridge-verisure/node_modules/verisure/index.js:27:17)
at Request.self.callback (/usr/local/lib/node_modules/homebridge-verisure/node_modules/request/request.js:188:22)
at emitTwo (events.js:106:13)
at Request.emit (events.js:191:7)
at Request. (/usr/local/lib/node_modules/homebridge-verisure/node_modules/request/request.js:1171:10)
at emitOne (events.js:96:13)
at Request.emit (events.js:188:7)
at IncomingMessage. (/usr/local/lib/node_modules/homebridge-verisure/node_modules/request/request.js:1091:12)`

When this happens, Homebridge crashes.

**Comment by siavashvj** [comment]
This also happen from the Home app, and if an autoamtion is set to lock/unlock if the lock is already locked/unlocked.

**Comment by kodsnutten** [comment]
Thanks @siavashvj !
I think the problem has been resolved now. The root cause was that the callback-function was called twice - which was fixed with [this commit](https://github.com/kodsnutten/homebridge-verisure/commit/fbb68e0db78bb43e6aea1461eca6ae6aa05fee17). I also [added errorhandling](https://github.com/kodsnutten/homebridge-verisure/commit/9ed661dd67afdc268ac2480101c7caf67c01be4c) for the verisure-api-specific error, when you try to set a already set state.

Another possible solution for the latter problem could be to first read the current state from the lock (from the Verisure API-endpoint), and then ONLY if "current state" differs from the "target state", execute the API-call to lock/unlock (this is a little bit slower, as it would add a API roundtrip to every lock/unlock-operation)


Anyway. You should be able to test the updated code by issuing `npm update -g homebridge-verisure` if you previously installed from this GIT-repository (using the command `npm install -g git+https://github.com/kodsnutten/homebridge-verisure.git`)

**Comment by siavashvj** [comment]
**Edit:** _I remember I installed this plugin by typing `npm install -g homebridge-verisure`, so maybe the update string below didn't work?_

**I did an update (I think) by typing `npm update -g homebridge-verisure` (didn't get a "success" message or anything, but I'm assuming it updated):**

> SMC-Mac-Mini:~ SMC$ npm update -g homebridge-verisure
> SMC-Mac-Mini:~ SMC$ 


**I then tried locking an already locked door via Siri, and got the same error:** 

> [7/15/2017, 12:17:30 AM] [Verisure] **** Response: Doorlockstate: {"statusCode":400,"body":"{\"errorGroup\":\"BAD_REQUEST\",\"errorCode\":\"VAL_TYPE_MISMATCH\",\"errorMessage\":\"Expected Long but was String (value=undefined). Invalid value for parameter transactionId\"}","headers":{"date":"Fri, 14 Jul 2017 22:17:30 GMT","content-type":"application/json;charset=UTF-8","connection":"close","transfer-encoding":"chunked"},"request":{"uri":{"protocol":"https:","slashes":true,"auth":null,"host":"e-api01.verisure.com","port":443,"hostname":"e-api01.verisure.com","hash":null,"search":null,"query":null,"pathname":"/xbn/2/installation/70775734374/doorlockstate/change/result/undefined","path":"/xbn/2/installation/70775734374/doorlockstate/change/result/undefined","href":"https://e-api01.verisure.com/xbn/2/installation/70775734374/doorlockstate/change/result/undefined"},"method":"GET","headers":{"Cookie":"vid=9h621619oc79mj3vnr45r8rjc8","Accept":"application/json, text/javascript, */*; q=0.01","Host":"e-api01.verisure.com"}}}
> /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/once.js:12
>       throw new Error("This callback function has already been called by someone else; it can only be called one time.");
>       ^
> 
> Error: This callback function has already been called by someone else; it can only be called one time.
>     at /usr/local/lib/node_modules/homebridge/node_modules/hap-nodejs/lib/util/once.js:12:13
>     at Object.<anonymous> (/usr/local/lib/node_modules/homebridge-verisure/index.js:293:11)
>     at Request._callback (/usr/local/lib/node_modules/homebridge-verisure/node_modules/verisure/index.js:27:17)
>     at Request.self.callback (/usr/local/lib/node_modules/homebridge-verisure/node_modules/request/request.js:188:22)
>     at emitTwo (events.js:106:13)
>     at Request.emit (events.js:191:7)
>     at Request.<anonymous> (/usr/local/lib/node_modules/homebridge-verisure/node_modules/request/request.js:1171:10)
>     at emitOne (events.js:96:13)
>     at Request.emit (events.js:188:7)
>     at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/homebridge-verisure/node_modules/request/request.js:1091:12)
>     at IncomingMessage.g (events.js:292:16)
>     at emitNone (events.js:91:20)
>     at IncomingMessage.emit (events.js:185:7)
>     at endReadableNT (_stream_readable.js:974:12)
>     at _combinedTickCallback (internal/process/next_tick.js:80:11)
>     at process._tickCallback (internal/process/next_tick.js:104:9)

**Comment by kodsnutten** [comment]
Hey @siavashvj ! It probably didn't get updated. If you just installed it using the command `npm install -g homebridge-verisure`, then you're not using the updated plugin (not even the Yale Doorman code).
I'd suggest you run:
`sudo npm uninstall -g homebridge-verisure`
and then
`sudo npm install -g git+https://github.com/kodsnutten/homebridge-verisure.git`

This will uninstall the "official" homebridge-verisure plugin (which does not currently have the Yale Doorman code), and then install the plugin from my repository.

When you have done this, you also need to restart homebridge!

**Comment by siavashvj** [comment]
Hmm, so if i'd installed via `npm install -g homebridge-verisure`I shouldn't have been able to control my Doorman? In that case, I must have installed via the GitHub repository, since I am able to control my Doorman. 

However, I uninstalled the homebridge-verisure and tried reinstalling it with `sudo npm install -g git+https://github.com/kodsnutten/homebridge-verisure.git` and got this error: 

> SMC-Mac-Mini:~ SMC$ sudo npm install -g git+https://github.com/kodsnutten/homebridge-verisure.git
> npm ERR! code 1
> npm ERR! Command failed: /usr/bin/git clone --depth=1 -q -b master https://github.com/kodsnutten/homebridge-verisure.git /Users/SMC/.npm/_cacache/tmp/git-clone-ca17eae2
> npm ERR! /Users/SMC/.npm/_cacache/tmp/git-clone-ca17eae2/.git: Permission denied
> npm ERR! 
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/SMC/.npm/_logs/2017-07-15T20_42_41_782Z-debug.log

**Comment by kodsnutten** [comment]
Woopsie... I haven't tried to install it on a MAC this way... I'm using Raspberry Pi to run my home automation - and that is Linux where "sudo" makes you get super user access (and is required in my system, if I want to install it with the "-g" parameter)... It seems that the root-user of your system does not have access to the directory `/Users/SMC/.npm/_cacache/tmp/git-clone-ca17eae2/.git` (or it simply can't create the direcory).

What if you simply try `npm install -g git+https://github.com/kodsnutten/homebridge-verisure.git`?

**Comment by siavashvj** [comment]
Ah :) 

No, no luck. Got more or less the same error, but a slightly different folder number for the git-clone-XXXXXXX-folder: 

> npm ERR! code 1
> npm ERR! Command failed: /usr/bin/git clone --depth=1 -q -b master https://github.com/kodsnutten/homebridge-verisure.git /Users/SMC/.npm/_cacache/tmp/git-clone-ea3026cb
> npm ERR! /Users/SMC/.npm/_cacache/tmp/**git-clone-ea3026cb**/.git: Permission denied
> npm ERR! 
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/SMC/.npm/_logs/2017-07-15T21_00_38_615Z-debug.log
> SMC-Mac-Mini:~ SMC$ 

When I navigate to `/Users/SMC/.npm/_cacache/tmp/git-clone-ea3026cb/.git` I only come to `/Users/SMC/.npm/_cacache/tmp` which is an empty folder.

**Comment by kodsnutten** [comment]
Bummer! Just tried it on my Macbook Pro, and it failed in the same way as for you...
It seems that npm on mac can't use direct github repos...

What you could do is to re install the homebridge-verisure plugin from NPM with the command:
`npm install -g homebridge-verisure`

and then replace the file `/usr/local/lib/node_modules/homebridge-verisure/index.js` with the [index.js](https://raw.githubusercontent.com/kodsnutten/homebridge-verisure/master/index.js) from my repository on github.

**Comment by siavashvj** [comment]
I don't know how, but this just happened: 

> SMC-Mac-Mini:~ SMC$ sudo npm install -g git+https://github.com/kodsnutten/homebridge-verisure.git
> homebridge-verisure@1.5.0
> added 56 packages in 3.713s

So i am up to date, however, my Yale Doorman now shows up as "not supported". It shows up, but I can't do anything with it since Home.app says it is not a supported device.

**Comment by kodsnutten** [comment]
Somehow I still think that you're using the "official" version of _homebridge-verisure_!

Could you please run this command, to confirm that you're using the correct plugin:
`grep Yale /usr/local/lib/node_modules/homebridge-verisure/index.js`

It should print one line:
`  'DOORLOCK': 'Yale Doorman',`

**Comment by siavashvj** [comment]
Never mind, I managed to reinstall it and it's working as it should again. 

Now I am on the latest version, with your fix for the already locked/unlocked crash, and after a few minutes testing the fix seems to be holding up smoothly! 

I've tried
+ locking an already locked door via Siri
+ unlocking an already unlocked door via Siri
+ The two above scenarios via automation/schedule

... and it's working! No more crashing! 

Good job, and thank you! The This should definitely be pushed to the master homebridge-verisure!
